File: packages/kotlin/src/main/kotlin/me/really/jose/Domain.kt
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

package me.really.jose

public enum class ReallyMeJoseSignatureAlgorithm { ED_DSA, ES256 }

public enum class ReallyMeJoseJweKeyManagementAlgorithm { DIRECT, ECDH_ES_P256, ECDH_ES_P384, ECDH_ES_P521 }

public enum class ReallyMeJoseJweContentEncryptionAlgorithm { A128_GCM, A192_GCM, A256_GCM }

public class ReallyMeJoseJwtHeaderPolicy(
    public val allowMissingType: Boolean,
    public val allowEmbeddedKeyHeader: Boolean,
    acceptedTypeValues: List<String>,
) {
    public val acceptedTypeValues: List<String> = acceptedTypeValues.toList()
}

public class ReallyMeJoseJwtTemporalPolicy(
    public val requireExpiration: Boolean,
    public val requireNotBefore: Boolean,
    public val requireIssuedAt: Boolean,
    public val clockSkewSeconds: Long,
    public val maximumFutureIssuedAtSkewSeconds: Long,
    public val nowUnix: Long,
    public val expectedAudience: String,
    public val expectedIssuer: String? = null,
    public val expectedSubject: String? = null,
)

/** Header policy keeps sensitive party information in wipeable arrays. */
public class ReallyMeJoseJweHeaderPolicy(
    public val requireKeyIdentifier: Boolean = false,
    public val expectedKeyIdentifier: String? = null,
    public val expectedType: String? = null,
    public val expectedContentType: String? = null,
    expectedAgreementPartyUInfo: ByteArray? = null,
    expectedAgreementPartyVInfo: ByteArray? = null,
) : AutoCloseable {
    private val ownedAgreementPartyUInfo: ByteArray? = expectedAgreementPartyUInfo?.copyOf()
    private val ownedAgreementPartyVInfo: ByteArray? = expectedAgreementPartyVInfo?.copyOf()

    internal fun agreementPartyUInfoCopy(): ByteArray? = ownedAgreementPartyUInfo?.copyOf()

    internal fun agreementPartyVInfoCopy(): ByteArray? = ownedAgreementPartyVInfo?.copyOf()

    /** Clears the policy's owned party-information copies. */
    override fun close() {
        ownedAgreementPartyUInfo?.fill(0)
        ownedAgreementPartyVInfo?.fill(0)
    }
}

File: packages/swift/Sources/ReallyMeJOSE/Domain.swift
// SPDX-FileCopyrightText: Copyright © 2026 ReallyMe LLC. All rights reserved
//
// SPDX-License-Identifier: Apache-2.0

public enum ReallyMeJOSESignatureAlgorithm: Sendable, Equatable {
  case edDSA
  case es256
}

public enum ReallyMeJOSEJWEKeyManagementAlgorithm: Sendable, Equatable {
  case direct
  case ecdhESP256
  case ecdhESP384
  case ecdhESP521
}

public enum ReallyMeJOSEJWEContentEncryptionAlgorithm: Sendable, Equatable {
  case a128GCM
  case a192GCM
  case a256GCM
}

public struct ReallyMeJOSEJWTHeaderPolicy: Sendable, Equatable {
  public let allowMissingTyp: Bool
  public let allowEmbeddedKeyHeader: Bool
  public let acceptedTypValues: [String]

  public init(
    allowMissingTyp: Bool,
    allowEmbeddedKeyHeader: Bool,
    acceptedTypValues: [String]
  ) {
    self.allowMissingTyp = allowMissingTyp
    self.allowEmbeddedKeyHeader = allowEmbeddedKeyHeader
    self.acceptedTypValues = acceptedTypValues
  }
}

public struct ReallyMeJOSEJWTTemporalPolicy: Sendable, Equatable {
  public let requireExpiration: Bool
  public let requireNotBefore: Bool
  public let requireIssuedAt: Bool
  public let clockSkewSeconds: UInt64
  public let maximumFutureIssuedAtSkewSeconds: UInt64
  public let nowUnix: UInt64
  public let expectedAudience: String
  public let expectedIssuer: String?
  public let expectedSubject: String?

  public init(
    requireExpiration: Bool,
    requireNotBefore: Bool,
    requireIssuedAt: Bool,
    clockSkewSeconds: UInt64,
    maximumFutureIssuedAtSkewSeconds: UInt64,
    nowUnix: UInt64,
    expectedAudience: String,
    expectedIssuer: String? = nil,
    expectedSubject: String? = nil
  ) {
    self.requireExpiration = requireExpiration
    self.requireNotBefore = requireNotBefore
    self.requireIssuedAt = requireIssuedAt
    self.clockSkewSeconds = clockSkewSeconds
    self.maximumFutureIssuedAtSkewSeconds = maximumFutureIssuedAtSkewSeconds
    self.nowUnix = nowUnix
    self.expectedAudience = expectedAudience
    self.expectedIssuer = expectedIssuer
    self.expectedSubject = expectedSubject
  }
}

public struct ReallyMeJOSEJWEHeaderPolicy: Sendable, Equatable {
  public let requireKeyIdentifier: Bool
  public let expectedKeyIdentifier: String?
  public let expectedType: String?
  public let expectedContentType: String?
  public let expectedAgreementPartyUInfo: [UInt8]?
  public let expectedAgreementPartyVInfo: [UInt8]?

  public init(
    requireKeyIdentifier: Bool = false,
    expectedKeyIdentifier: String? = nil,
    expectedType: String? = nil,
    expectedContentType: String? = nil,
    expectedAgreementPartyUInfo: [UInt8]? = nil,
    expectedAgreementPartyVInfo: [UInt8]? = nil
  ) {
    self.requireKeyIdentifier = requireKeyIdentifier
    self.expectedKeyIdentifier = expectedKeyIdentifier
    self.expectedType = expectedType
    self.expectedContentType = expectedContentType
    self.expectedAgreementPartyUInfo = expectedAgreementPartyUInfo
    self.expectedAgreementPartyVInfo = expectedAgreementPartyVInfo
  }
}
